### PR TITLE
Bugfix/custom extra vars env refresh

### PR DESCRIPTION
**Pull request by radurentea**
## Description

After the extension refactor (#1868) the ESP-IDF environment is expanded once when a setup is loaded and every command reads that snapshot from the `CURRENT_IDF_CONFIGURATION` store. Two regressions came with that:

- The `idf.customExtraVars` change handler wrote every custom var raw into the snapshot, skipping the expansion that derives `OPENOCD_SCRIPTS` from the OpenOCD binary in PATH. Set Target and OpenOCD board selection both write that setting, so a stale `OPENOCD_SCRIPTS` left in `settings.json` by an older install was re-injected on every target change. On `release/v2.3.0` JTAG flash then fails with `OpenOCD config file(s) not found under OPENOCD_SCRIPTS`. Vars removed from the setting also never left the snapshot.
- Selecting, switching, deleting or failing to parse a project configuration never refreshed the snapshot, so a preset's `environment` block and `IDF_PRESET` only took effect after a window reload.

A third, older problem surfaced at the same time: `idf.customExtraVars` always overrides the setup env, so a stale `ESP_ROM_ELF_DIR` from an old install went into the build and, now that debugging sources the generated `gdbinit/symbols` file, debug sessions fail on a ROM ELF that does not exist.

This PR:

1. Returns a copy from `getCurrentIdfConfiguration` so callers that edit their own env no longer mutate the stored snapshot.
2. Keeps the raw setup environment in the store and re-expands it whenever `idf.customExtraVars` changes, so there is a single expansion path.
3. Re-expands on every project configuration selection change, and clears the previous folder's selected preset whenever the selected workspace folder changes, so a name kept after a preset parse failure cannot leak into `IDF_PRESET` or select a same-named preset in the next folder. Activation on the same folder keeps the remembered selection.
4. Skips custom vars whose value is an absolute path that no longer exists while the setup provides the same variable, reports them as warnings in Doctor and shows a notification that explains which paths are used and can remove the obsolete entries. Removal only touches the settings scope that holds the stale value, so a same-named valid value in the user or workspace scope is kept.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Steps to test this pull request

Use a `settings.json` that mimics an old install, for example:

```json
"idf.customExtraVars": {
  "OPENOCD_SCRIPTS": "/path/to/.espressif/tools/openocd-esp32/v0.12.0-esp32-20240318/openocd-esp32/share/openocd/scripts",
  "ESP_ROM_ELF_DIR": "/path/to/.espressif/tools/esp-rom-elfs/20230320/",
  "ADF_PATH": "/path/to/esp-adf",
  "IDF_TARGET": "esp32c6"
}
```

where the two tool paths do not exist and ADF_PATH does. To check the scoped removal, also add "idf.customExtraVars": { "OPENOCD_SCRIPTS": "<an existing directory>" } to your user settings.

1. Reload the window with an ESP-IDF 6.x setup selected.
2. Run ESP-IDF: Set Espressif Device Target and pick the connected board.
3. Build, then ESP-IDF: Flash your project with flash method JTAG.
4. Start a debug session.
5. Run ESP-IDF: Doctor Command.
6. Click Remove obsolete entries on the notification shown after the reload.

- Expected behaviour:
  - After step 1 a warning notification names OPENOCD_SCRIPTS, ESP_ROM_ELF_DIR and says the setup paths are used instead.
  - Steps 3 and 4 succeed without editing settings.
  - Doctor shows two warn lines for those keys and lists the current setup's OPENOCD_SCRIPTS and ESP_ROM_ELF_DIR under "ESP-IDF Project Setup Variables".
  - Step 6 removes only those two keys from the folder settings; ADF_PATH and IDF_TARGET stay, and the OPENOCD_SCRIPTS entry in user settings is untouched.
- Expected output: ⚡️ Flashed Successfully (JTAG) in the Flash output channel and a debug session that stops at app_main.

How has this been tested?

As described above. Also tested multiple project configurations (preset environment blocks and IDF_PRESET follow the selection without a reload) and multi-root workspaces (two copies of the ESP-IDF multi_config example with identical preset names: after breaking CMakePresets.json in one folder and switching to the other, no preset is selected there and Doctor shows no IDF_PRESET; the remembered selection survives a reload on the same folder).

Test Configuration:

- ESP-IDF Version: v6.1
- OS (Windows,Linux and macOS): Windows

Dependent components impacted by this PR:

- ESP-IDF environment store (CURRENT_IDF_CONFIGURATION, new CURRENT_IDF_SETUP_ENV) and every command that reads it
- idf.customExtraVars change handling
- Project configuration selection (CMakePresets)
- Workspace folder switching
- OpenOCD launch, JTAG flash, debug
- Doctor command report

Checklist

- [x] PR Self Reviewed
- [x] Applied Code formatting
- [ ] Added Documentation
- [x] Added Unit Test
- [ ] Verified on all platforms - Windows,Linux and macOS

## Summary by CodeRabbit

- **New Features**
  - Detects stale custom environment variables whose configured paths no longer exist.
  - Reports stale variables in configuration analysis and explains when the ESP-IDF setup value is used.
  - Provides actions to remove obsolete entries or open the relevant setting.

- **Bug Fixes**
  - Refreshes ESP-IDF environment settings when configurations are selected, forgotten, or suspended.
  - Reapplies environment settings when custom variables or OpenOCD paths change.
  - Preserves environment configuration when callers modify retrieved values.
  - Cleans up project configuration indicators only when switching workspace folders.

### CODE REVIEW

File: src/common/configurationChange.ts [hunk 1]
```diff
@@ -93,9 +102,9 @@ export function registerOnDidChangeConfiguration(context: ExtensionContext) {
               customExtraVars[envVar],
               { applyAtProcessCreation: true }
             );
-            updateCurrentIdfEnvVar(envVar, customExtraVars[envVar]);
           }
         }
+        await refreshCurrentIdfConfiguration(prevWorkspaceFolder);
```
[the remark this answers is not in the document]
Comment (radurentea) [reply src/common/configurationChange.ts:95]: Correct observation, but filtering here would keep a write that should not exist. The whole environmentVariableCollection.replace loop is removed in #1949, because pushing idf.customExtraVars into the collection is what leaks stale values into terminals (#1883). This branch will be rebased once #1949 is merged, at which point this handler only runs the refresh.

**Comment by brianignacio5** [comment]
Thank you for the PR.

I agree there is value in determining stale setting as you are doing I have 2 additional suggestions to avoid conflict of Env Variables from idf.customExtraVars with ESP-IDF setup.

1) I propose to have OPENOCD_SCRIPTS in `expandEnvVariablesForIdfSetup` to be ignored from idf.customExtraVars (see prepareEnv.ts line 100 for PATH) and have OPENOCD_SCRIPTS to be resolved using `OpenOCDManager.getOpenOcdPath` as it is done in lines 115-133 of prepareEnv.ts

Although looking at the code of expandEnvVariablesForIdfSetup, even if OPENOCD_SCRIPTS is defined in idf.customExtraVars, shouldn't be always override by OPENOCD_SCRIPTS from current openOCD in PATH (115-133 of prepareEnv.ts) ? 

2) We should also ignores ESP_ROM_ELF_DIR from idf.customExtraVars (see prepareEnv.ts line 100 for PATH).

Other changes to keep update of IDF_PRESET and actually having both IDF_ENV and IDF_CURRENT_SETUP sounds good to me. In general I agree with idea, I just think that OpenOCD path and OPENOCD_SCRIPTS should always come from IDF_ENV unless user sets a `"idf.customOpenOCDPath"` which should also be part of Doctor Command if defined.

**Comment by radurentea** [comment]
1. Agreed. You read it right: the block after the custom-vars loop already overwrites `OPENOCD_SCRIPTS` whenever an OpenOCD binary is found via `idf.customOpenOCDPath` or PATH and its scripts folder exists, so a custom value only survives when no binary is found at all. I'll skip the key explicitly like `PATH` so that edge case falls back to the setup value instead, add a test, and fix the `idf.customExtraVars` doc entry, which currently uses `OPENOCD_SCRIPTS` as its example.

2. On `ESP_ROM_ELF_DIR` I'd suggest keeping the generic rule rather than ignoring it outright. Unlike OpenOCD there is no other setting to override it, and ESP-IDF itself honours a user-exported value, so it's the only way to point the debugger at ROM ELFs that aren't in the tools install yet (newer chip revision, local `esp-rom-elfs` build). The stale check already drops it when the path doesn't exist, which is the case that broke debugging, and it applies to every `tools.json` export var without a hard-coded list. If you'd still rather ignore it, it's a one-line addition to the same skip, happy to do it.

`idf.customOpenOCDPath` is already printed in the Doctor report when defined, with an accessibility check in the summary. I can also add the resolved OpenOCD binary path so it's visible which one is used when the setting is empty, if you think that's useful.

**Review by AndriiFilippov — APPROVED** [review]
@radurentea  hi ! 
Verified on Windows 11 / Mac arm64

Stale `OPENOCD_SCRIPTS` and `ESP_ROM_ELF_DIR` are detected and ignored, with ESP-IDF setup values used as fallback ✔️ 
Valid custom OpenOCD is used correctly ✔️ 
Invalid  `idf.customOpenOCDPath` is detected and ESP-IDF-managed OpenOCD is used as fallback ✔️ 
JTAG/OpenOCD and GDB debugging were verified successfully up to the app_main breakpoint. ✔️ 
LGTM 👍

**Review by brianignacio5 — APPROVED** [review]
LGTM
